File: legacy/server-node/src/routerstore.js
/**
 * Almacén de routers configurados (tabla `routers` de SQLite) + bootstrap:
 *
 *  1. Si la tabla tiene filas → es la fuente de verdad.
 *  2. Si está vacía y hay ROUTERS_JSON en el entorno → se siembra desde ahí.
 *  3. Si sigue vacía y NO estamos en demo forzado → autodetección de la
 *     puerta de enlace: `ip route show default` en el propio servidor y
 *     sondeo SSH (`ubus call system board`) para averiguar modelo y tipo
 *     (GL.iNet vs OpenWrt genérico). El resto de routers se añaden a mano
 *     desde Ajustes (CRUD /api/config/routers).
 *
 * La tabla NACE VACÍA al importar el código: no hay datos de semilla.
 */
import { execFile } from 'node:child_process'
import { sshBaseArgs } from './sshkey.js'

const PROBE_TIMEOUT_MS = 4000

export function listRouters(db) {
  return db
    .prepare('SELECT id, name, host, type, is_gateway, created_at FROM routers ORDER BY is_gateway DESC, created_at ASC')
    .all()
    .map((r) => ({ ...r, is_gateway: r.is_gateway === 1 }))
}

function slugify(text) {
  return String(text)
    .toLowerCase()
    .normalize('NFD')
    .replace(/[̀-ͯ]/g, '')
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '')
    .slice(0, 32)
}

/** Genera un id único a partir del nombre o del host. */
function uniqueId(db, name, host) {
  const base = slugify(name || '') || `router-${String(host).split('.').pop()}`
  let id = base || 'router'
  let n = 2
  while (db.prepare('SELECT 1 FROM routers WHERE id = ?').get(id)) {
    id = `${base}-${n++}`
  }
  return id
}

/**
 * Inserta un router y devuelve la fila creada.
 * Si `isGateway`, el resto pierde el flag (solo un gateway).
 */
export function addRouter(db, { name, host, type = 'openwrt', isGateway = false }) {
  const id = uniqueId(db, name, host)
  const now = Date.now()
  const tx = db.transaction(() => {
    if (isGateway) db.prepare('UPDATE routers SET is_gateway = 0').run()
    db.prepare('INSERT INTO routers (id, name, host, type, is_gateway, created_at) VALUES (?, ?, ?, ?, ?, ?)').run(
      id,
      name || host,
      host,
      type,
      isGateway ? 1 : 0,
      now,
    )
  })
  tx()
  return listRouters(db).find((r) => r.id === id)
}

export function removeRouter(db, id) {
  return db.prepare('DELETE FROM routers WHERE id = ?').run(id).changes > 0
}

// ---------------------------------------------------------------------------
// Autodetección de la puerta de enlace
// ---------------------------------------------------------------------------

function exec(cmd, args, timeoutMs = PROBE_TIMEOUT_MS) {
  return new Promise((resolve, reject) => {
    execFile(cmd, args, { timeout: timeoutMs, maxBuffer: 1024 * 1024 }, (err, stdout, stderr) => {
      if (err) return reject(new Error(`${cmd}: ${err.message} ${stderr || ''}`.trim()))
      resolve(stdout)
    })
  })
}

/** IP de la puerta de enlace por defecto del servidor (null si no se sabe). */
export async function detectGatewayIp() {
  try {
    const out = await exec('ip', ['route', 'show', 'default'])
    const m = /default via (\d+\.\d+\.\d+\.\d+)/.exec(out)
    return m ? m[1] : null
  } catch {
    return null
  }
}

/** Sondea el modelo del gateway por SSH (ubus system board). */
async function probeGatewayModel(host, sshKeyPath) {
  try {
    const out = await exec(
      'ssh',
      [
        ...sshBaseArgs(sshKeyPath),
        '-o', 'ConnectTimeout=3',
        '-o', 'ControlMaster=no',
        `root@${host}`,
        'ubus call system board',
      ],
      PROBE_TIMEOUT_MS + 1000,
    )
    const board = JSON.parse(out)
    const model = board?.model || ''
    const isGlinet = /GL[.-]?iNet|GL-[A-Z]/i.test(model) || /glinet/i.test(board?.release?.distribution || '')
    return { model, type: isGlinet ? 'glinet' : 'openwrt' }
  } catch {
    return { model: '', type: 'openwrt' }
  }
}

/**
 * Bootstrap de la tabla routers. Devuelve la lista final configurada.
 * @param {object} db - handle better-sqlite3
 * @param {object} config - config cargada (routers de ROUTERS_JSON, sshKeyPath, demoMode)
 * @param {(msg: string) => void} log
 */
export async function ensureInitialRouters(db, config, log = console.log) {
  const existing = listRouters(db)
  if (existing.length > 0) return existing

  // 1. Semilla desde ROUTERS_JSON (primer arranque)
  if (config.routers.length > 0) {
    for (const [i, r] of config.routers.entries()) {
      addRouter(db, { name: r.name, host: r.host, type: r.type, isGateway: r.type === 'glinet' || i === 0 })
    }
    log(`[netpulse] routers sembrados desde ROUTERS_JSON (${config.routers.length})`)
    return listRouters(db)
  }

  // 2. Autodetección de la puerta de enlace (nunca en demo forzado)
  if (config.demoMode) return []

  const gwIp = await detectGatewayIp()
  if (!gwIp) {
    log('[netpulse] sin puerta de enlace detectada; añade routers desde Ajustes')
    return []
  }
  const probe = await probeGatewayModel(gwIp, config.sshKeyPath)
  const created = addRouter(db, {
    name: probe.model || 'Gateway',
    host: gwIp,
    type: probe.type,
    isGateway: true,
  })
  log(
    `[netpulse] gateway autodetectado: ${created.host} (${probe.model || 'modelo desconocido'}, tipo ${created.type}). El resto se añade desde Ajustes.`,
  )
  return listRouters(db)
}

File: server-go/internal/routerstore/routerstore.go
// Package routerstore — almacén de routers configurados (tabla `routers` de
// SQLite) + bootstrap (paridad src/routerstore.js, SPEC §8.2):
//  1. Si la tabla tiene filas → es la fuente de verdad.
//  2. Si está vacía y hay ROUTERS_JSON → se siembra.
//  3. Si sigue vacía y NO es demo → autodetección del gateway (ip route +
//     sondeo SSH `ubus call system board`).
package routerstore

import (
	"database/sql"
	"encoding/json"
	"log"
	"os/exec"
	"regexp"
	"strings"
	"time"
	"unicode"

	"github.com/gnacho/netpulse/server-go/internal/adapters"
	"github.com/gnacho/netpulse/server-go/internal/config"
	"github.com/gnacho/netpulse/server-go/internal/sshkey"
	"golang.org/x/text/unicode/norm"
)

const probeTimeout = 4 * time.Second

// ListRouters devuelve la tabla routers ordenada is_gateway DESC,
// created_at ASC, con is_gateway como booleano.
func ListRouters(db *sql.DB) []adapters.RouterConfig {
	rows, err := db.Query("SELECT id, name, host, type, is_gateway, agent_only, created_at FROM routers ORDER BY is_gateway DESC, created_at ASC")
	if err != nil {
		return []adapters.RouterConfig{}
	}
	defer rows.Close()
	out := []adapters.RouterConfig{}
	for rows.Next() {
		var r adapters.RouterConfig
		var gw, ao int
		var name sql.NullString
		if err := rows.Scan(&r.ID, &name, &r.Host, &r.Type, &gw, &ao, &r.CreatedAt); err != nil {
			continue
		}
		r.Name = name.String
		r.IsGateway = gw == 1
		r.AgentOnly = ao == 1
		out = append(out, r)
	}
	return out
}

var slugNonAlnum = regexp.MustCompile(`[^a-z0-9]+`)

// slugify replica el slug JS: minúsculas, NFD sin diacríticos,
// [^a-z0-9]+ → '-', trim guiones, máx 32.
func slugify(text string) string {
	s := strings.ToLower(text)
	// NFD + eliminar marcas combinantes (U+0300–U+036F)
	var b strings.Builder
	for _, r := range norm.NFD.String(s) {
		if r >= 0x0300 && r <= 0x036F {
			continue
		}
		if !unicode.Is(unicode.Mn, r) {
			b.WriteRune(r)
		}
	}
	s = slugNonAlnum.ReplaceAllString(b.String(), "-")
	s = strings.Trim(s, "-")
	if len(s) > 32 {
		s = s[:32]
	}
	return s
}

// uniqueId genera un id único a partir del nombre o del host.
func uniqueId(db *sql.DB, name, host string) string {
	base := slugify(name)
	if base == "" {
		parts := strings.Split(host, ".")
		base = "router-" + parts[len(parts)-1]
	}
	if base == "" {
		base = "router"
	}
	id := base
	n := 2
	for {
		var one int
		err := db.QueryRow("SELECT 1 FROM routers WHERE id = ?", id).Scan(&one)
		if err == sql.ErrNoRows {
			return id
		}
		if err != nil {
			return id
		}
		id = base + "-" + itoa(n)
		n++
	}
}

func itoa(n int) string {
	if n == 0 {
		return "0"
	}
	var buf [8]byte
	i := len(buf)
	for n > 0 {
		i--
		buf[i] = byte('0' + n%10)
		n /= 10
	}
	return string(buf[i:])
}

// AddInput son los datos de alta de un router.
type AddInput struct {
	Name      string
	Host      string
	Type      string // default "openwrt"
	IsGateway bool
	AgentOnly bool
}

// AddRouter inserta un router (si IsGateway, el resto pierde el flag —
// transacción, un solo gateway) y devuelve la fila creada.
func AddRouter(db *sql.DB, in AddInput) (adapters.RouterConfig, error) {
	if in.Type == "" {
		in.Type = "openwrt"
	}
	id := uniqueId(db, in.Name, in.Host)
	now := time.Now().UnixMilli()
	name := in.Name
	if name == "" {
		name = in.Host
	}
	tx, err := db.Begin()
	if err != nil {
		return adapters.RouterConfig{}, err
	}
	defer tx.Rollback()
	if in.IsGateway {
		if _, err := tx.Exec("UPDATE routers SET is_gateway = 0"); err != nil {
			return adapters.RouterConfig{}, err
		}
	}
	gw := 0
	if in.IsGateway {
		gw = 1
	}
	ao := 0
	if in.AgentOnly {
		ao = 1
	}
	if _, err := tx.Exec(
		"INSERT INTO routers (id, name, host, type, is_gateway, agent_only, created_at) VALUES (?, ?, ?, ?, ?, ?, ?)",
		id, name, in.Host, in.Type, gw, ao, now,
	); err != nil {
		return adapters.RouterConfig{}, err
	}
	if err := tx.Commit(); err != nil {
		return adapters.RouterConfig{}, err
	}
	for _, r := range ListRouters(db) {
		if r.ID == id {
			return r, nil
		}
	}
	return adapters.RouterConfig{}, sql.ErrNoRows
}

// RemoveRouter borra por id; true si borró (changes > 0).
func RemoveRouter(db *sql.DB, id string) bool {
	res, err := db.Exec("DELETE FROM routers WHERE id = ?", id)
	if err != nil {
		return false
	}
	n, _ := res.RowsAffected()
	return n > 0
}

// ---------------------------------------------------------------------------
// Autodetección de la puerta de enlace
// ---------------------------------------------------------------------------

var defaultViaRe = regexp.MustCompile(`default via (\d+\.\d+\.\d+\.\d+)`)

// DetectGatewayIP devuelve la IP de la puerta de enlace por defecto del
// servidor ("" si no se sabe).
func DetectGatewayIP() string {
	out, err := exec.Command("ip", "route", "show", "default").Output()
	if err != nil {
		return ""
	}
	if m := defaultViaRe.FindSubmatch(out); m != nil {
		return string(m[1])
	}
	return ""
}

var glinetModelRe = regexp.MustCompile(`(?i)GL[.-]?iNet|GL-[A-Z]`)
var glinetDistRe = regexp.MustCompile(`(?i)glinet`)

// probeGatewayModel sondea el modelo del gateway por SSH (ubus system board).
func probeGatewayModel(host, keyPath string) (model, typ string) {
	args := append(sshkey.BaseArgs(keyPath),
		"-o", "ConnectTimeout=3",
		"-o", "ControlMaster=no",
		"root@"+host,
		"ubus call system board",
	)
	cmd := exec.Command("ssh", args...)
	type result struct {
		out []byte
		err error
	}
	ch := make(chan result, 1)
	go func() {
		out, err := cmd.Output()
		ch <- result{out, err}
	}()
	select {
	case <-time.After(probeTimeout + time.Second):
		_ = cmd.Process.Kill()
		return "", "openwrt"
	case res := <-ch:
		if res.err != nil {
			return "", "openwrt"
		}
		var board struct {
			Model   string `json:"model"`
			Release struct {
				Distribution string `json:"distribution"`
			} `json:"release"`
		}
		if err := json.Unmarshal(res.out, &board); err != nil {
			return "", "openwrt"
		}
		typ = "openwrt"
		if glinetModelRe.MatchString(board.Model) || glinetDistRe.MatchString(board.Release.Distribution) {
			typ = "glinet"
		}
		return board.Model, typ
	}
}

// EnsureInitialRouters es el bootstrap de la tabla routers
// (routerstore.js:126-158). Devuelve la lista final configurada.
func EnsureInitialRouters(db *sql.DB, cfg *config.Config) []adapters.RouterConfig {
	existing := ListRouters(db)
	if len(existing) > 0 {
		return existing
	}

	// 1. Semilla desde ROUTERS_JSON (primer arranque)
	if len(cfg.Routers) > 0 {
		for i, r := range cfg.Routers {
			_, err := AddRouter(db, AddInput{
				Name: r.Name, Host: r.Host, Type: r.Type,
				IsGateway: r.Type == "glinet" || i == 0,
			})
			if err != nil {
				log.Printf("[netpulse] error sembrando router %s: %v", r.Host, err)
			}
		}
		log.Printf("[netpulse] routers sembrados desde ROUTERS_JSON (%d)", len(cfg.Routers))
		return ListRouters(db)
	}

	// 2. Autodetección de la puerta de enlace (nunca en demo forzado)
	if cfg.DemoMode {
		return []adapters.RouterConfig{}
	}
	gwIP := DetectGatewayIP()
	if gwIP == "" {
		log.Printf("[netpulse] sin puerta de enlace detectada; añade routers desde Ajustes")
		return []adapters.RouterConfig{}
	}
	model, typ := probeGatewayModel(gwIP, cfg.SSHKeyPath)
	name := model
	if name == "" {
		name = "Gateway"
	}
	created, err := AddRouter(db, AddInput{Name: name, Host: gwIP, Type: typ, IsGateway: true})
	if err != nil {
		log.Printf("[netpulse] error creando gateway autodetectado: %v", err)
		return ListRouters(db)
	}
	if model == "" {
		model = "modelo desconocido"
	}
	log.Printf("[netpulse] gateway autodetectado: %s (%s, tipo %s). El resto se añade desde Ajustes.",
		created.Host, model, created.Type)
	return ListRouters(db)
}
